File: 74-search-a-2d-matrix/74-search-a-2d-matrix.js
/**
 * @param {number[][]} matrix
 * @param {number} target
 * @return {boolean}
 */
var searchMatrix = function(matrix, target) {
    const m = matrix.length;
    const n = matrix[0].length;
    let start = 0, end = m*n-1;
    let mid = Math.round((end - start)/2);
    
    while(start <= end) {
        console.log(start + " " + end + " " + matrix[Math.trunc(mid/n)][Math.round(mid%n)]);
        if(matrix[Math.trunc(mid/n)][Math.round(mid%n)] == target) return true;
        
        if(matrix[Math.trunc(mid/n)][Math.round(mid%n)] > target) {
            end = mid - 1;
        } else {
            start = mid + 1;
        }
        
        mid = Math.round(start + (end-start)/2);
    }
    
    return false;
};

File: 74-search-a-2d-matrix/74-search-a-2d-matrix.cpp
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        int m = matrix.size();
        int n = matrix[0].size();
        
        int st = 0;
        int ed = m*n - 1;
        int md = (st+ed)/2;
        
        while(st <= ed) {
            // cout << st << " " << md << " " << ed << "\n";
            if(matrix[md/n][md%n] == target) return true;
            
            if(matrix[md/n][md%n] <= target) {
                st = md+1;
            } else {
                ed = md-1;
            }
            // cout << "here";
            md = (st+ed)/2;
        }
        
        return false;
    }
};